src/Agent.Worker/ContainerOperationProvider.cs

```diff
@@ -316,8 +316,30 @@ private async Task PullContainerAsync(IExecutionContext executionContext, Contai
                 ArgUtil.NotNullOrEmpty(username, nameof(username));
                 ArgUtil.NotNullOrEmpty(password, nameof(password));
 
-                int loginExitCode = await _dockerManger.DockerLogin(executionContext, registryServer, username, password);
-                if (loginExitCode != 0)
+
+                int retryCount = 0;
+                int loginExitCode = 0;
+
+                while (retryCount < 3)
+                {
+                    loginExitCode = await _dockerManger.DockerLogin(executionContext, registryServer, username, password);
+                    if (loginExitCode == 0)
+                    {
+                        break;
+                    }
+                    else
+                    {
+                        retryCount++;
+                        if (retryCount < 3)
+                        {
+                            var backOff = BackoffTimerHelper.GetRandomBackoff(TimeSpan.FromSeconds(1), TimeSpan.FromSeconds(10));
+                            executionContext.Warning($"Docker login failed with exit code {loginExitCode}, back off {backOff.TotalSeconds} seconds before retry.");
+                            await Task.Delay(backOff);
+                        }
+                    }
+                }
```

Comment:
Looks good. [suggestion/not blocking] It would be nice if the retry behavior was gated by a feature flag so it could be turned off if it causes problems in production w/o needing to roll back the agent.

[promoting this comment to a 'Conversation']